Auth: 비밀번호 찾기(재설정)

wch-os opened the pull request:
## Issues

- Resolves #398 
- Resolves #399 
- Resolves #400 

## Description

- 비밀번호 찾기(재설정) 기능 구현

- [POST] `email/password/reset/send`
  - 사용자 이메일로 비밀번호를 변경할 수 있는 링크 발송

- [POST] `email/password/reset`
  - 이메일 링크를 통해 비밀번호 변경

<br />

## Review Points

- 자유로운 리뷰 부탁드립니다.

<br />

## How Has This Been Tested?

- 포스트맨 API 테스트를 통해 이메일 전송과 비밀번호 재설정 기능 동작을 확인했습니다.

<br />

## Additional Notes

- 기능을 명확하게 구분하기 위해 사용자가 비로그인 상태에서 비밀번호를 잊었을 때 사용하는 기능인 **비밀번호 찾기(재설정)**, 사용자가 로그인 상태에서 마이페이지를 통해 진행되는 **비밀번호 변경** 기능을 별도의 이슈로 분리했습니다.
- 사용자의 안전한 비밀번호 설정을 위한, **비밀번호 히스토리 관리 기능**은 이후에 고도화를 진행하려 합니다.

wch-os committed:
feat(auth): implement password reset email usecase

- 비밀번호 재설정 클라이언트 구별을 위한 nonce 토큰 발급 및 redis 저장
- 비밀번호 재설정 이메일 전송 기능 정의

wch-os committed:
feat(auth): implement password reset email sending

- 비밀번호 재설정 이메일 발송 구현

wch-os committed:
feat(auth): implement password reset usecase

- 올바른 클라이언트의 비밀번호 재설정 요청인지 nonce 토큰을 비교하여 검증
- 새로운 비밀번호 암호화 및 DB에서 비밀번호 업데이트 기능 정의

wch-os committed:
feat(auth): implement `updatePasswordByEmail` in port adapter

wch-os committed:
feat(auth): add password reset button with link

wch-os committed:
feat(auth): define email password reset APIs in web layer

이메일 비밀번호 재설정 링크 전송
- POST: email/password/reset/send

이메일 비밀번호 재설정
- POST: email/password/reset

wch-os committed:
fix(auth): add createdAt and updatedAt to User domain

- 비밀번호 업데이트 과정에서 User 도메인에 createdAt, updatedAt 필드가 없어 에러가 발생함
- 누락된 필드를 추가하여 문제 해결

wch-os committed:
refactor(auth): rename `updatePasswordByEmail` to `updatePassword`

merge-simpson commented on services/auth/application/src/main/java/nettee/auth/port/MailSender.java:
💊 조금 더 구체적인 이름으로 수정하면 좋을 것 같습니다.

보니까 인증 관련 메일의 구체적인 사례들을 다루는 인터페이스로 보입니다!
다른 도메인의 기능을 포괄하지 않으므로 구체적인 네이밍을 제안드립니다.

merge-simpson commented on services/auth/application/src/main/java/nettee/auth/service/AuthCommandService.java:
```diff
@@ -199,6 +203,47 @@ public void withdraw(String userId, String refreshToken) {
+    @Override
+    public String sendPasswordResetEmail(String email) {
+        // 사용자 존재 여부 확인
+        authQueryRepositoryPort.findByEmail(email)
+                .orElseThrow(() -> new AuthException(AUTH_ACCOUNT_NOT_FOUND));
```
❓ 민감한 작업을 할 때는 command repository로부터 조회할 수도 있습니다. 어떤 게 나을까요?

한번 같이 고민해 주시면 좋을 것 같습니다.
우선순위도 높지 않고, 당장 문제가 되지는 않을 것 같습니다!

> 키워드: TOCTOU (Time-of-Check, Time-of-Use) 레이스

---

사용자가 '조회 요청'을 할 때는 query용 포트를 활용하고, 그 외 명령 요청에 대해서는 command중심적인 처리를 고려하면 조금 더 나을 수도 있습니다!

잘 고민해 주시면 좋을 것 같아요.

merge-simpson commented on services/auth/driving/web-mvc/src/main/java/nettee/auth/web/AuthCommandApi.java:
```diff
@@ -122,28 +122,27 @@ public ResponseEntity<String> sendEmailVerification(@RequestBody EmailVerifySend
     )
-    public void changePassword(@RequestBody PasswordVerifyRequest passwordVerifyRequest) {
-        // 비밀번호 변경 로직 구현
+    public ResponseEntity<String> sendPasswordResetEmail(@RequestBody PasswordForgotRequest request) {
+        String nonce = authSignUsecase.sendPasswordResetEmail(request.email());
+        return ResponseEntity.ok(nonce);
```
💊 응답 DTO 객체를 추가하면 어떨까요?

물론 문자열 반환만으로 충분한 시나리오지만, 프로젝트 내 작업 일관성을 위해서 DTO로 감싸서 반환하면 좋을 것 같습니다!

merge-simpson commented on services/auth/driving/web-mvc/src/main/java/nettee/auth/web/AuthCommandApi.java:
```diff
@@ -122,28 +122,27 @@ public ResponseEntity<String> sendEmailVerification(@RequestBody EmailVerifySend
     @Operation(
-            summary = "비밀번호 변경",
-            description = "사용자의 비밀번호를 변경합니다."
+            summary = "이메일 비밀번호 재설정",
+            description = "사용자가 이메일 링크를 통해 비밀번호를 변경합니다."
     )
-    public void resetPassword(@RequestBody PasswordResetRequest passwordResetRequest) {
-        // 비밀번호 재설정 링크 전송 로직 구현
+    public void resetPassword(@RequestBody PasswordResetRequest request) {
+        authSignUsecase.resetPassword(request.email(), request.newPassword(), request.nonce());
```
`void` 또는 `ResponseEntity<Void>` 응답 시, 수정이나 삭제에서 특별히 반환할 응답 데이터가 없을 때는, 다음 상태코드로 응답하면 더 좋아 보입니다:

```java
@ResponseStatus(HttpStatus.NO_CONTENT)
```

일반적인 실무 환경에서 민감한 차이를 야기하는 것은 아니지만, 만들 때 서로 발견해 줄 수 있으면 좋을 것 같아요.

merge-simpson commented on services/auth/driving/web-mvc/src/main/java/nettee/auth/web/dto/AuthCommandDto.java:
```diff
@@ -109,18 +109,39 @@ public record EmailVerifyRequest(
+            nonce = nonce.strip();
+
+            // 비밀번호 정규식 검증
+            final String PASSWORD_REGEX = "^[A-Za-z\\d !\"#$%&'()*+,\\-./:;<=>?@\\[\\]^_`{|}~]+$";
```
정책을 최대한 반영해 주신 것 같은데, 공백( )을 허용하는 것으로 보입니다!
띄어쓰기는 비밀번호에서 보통 허용하지 않는 것 같으니, 정책을 다시 살펴보면 좋을 것 같습니다!

merge-simpson reviewed (approved):
@wch-os 님, 빠른 작업 감사드립니다!

사소하게 수정하면 좋을 부분도 있었지만, 대부분 간단한 수정이거나 지금 단계에서 생략할 수 있는 것들이었습니다.

미리 approve를 해 둘 테니, 필요한 만큼 수정해서 병합하시면 좋을 것 같습니다!

wch-os committed:
refactor(auth): return nonce token via URL parameter for password reset

- 비밀번호 재설정 로직에서 nonce 토큰 반환 방식 수정 (response body(JSON) → url parameter)

- json 응답 방식일 시, 해당 응답을 받는 클라이언트에서 비밀번호 재설정이 일어나지 않는다.
비밀번호 재설정은 이메일에 포함된 링크를 통해 이루어지므로, nonce 토큰은 반드시 비밀번호 재설정 링크에 포함되어 반환되어야 한다.

wch-os committed:
refactor(auth): renames `MailSender` to `AuthMailSender`

wch-os committed:
refactor(auth): return 204 No Content for void responses

wch-os committed:
fix(auth): update password regex to disallow space characters

wch-os committed:
Merge remote-tracking branch 'origin/develop' into feature/auth-password-reset

merge-simpson commented on services/auth/driving/web-mvc/src/main/java/nettee/auth/web/AuthCommandApi.java:
```diff
@@ -122,28 +125,29 @@ public ResponseEntity<String> sendEmailVerification(@RequestBody EmailVerifySend
     )
-    public void changePassword(@RequestBody PasswordVerifyRequest passwordVerifyRequest) {
-        // 비밀번호 변경 로직 구현
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public ResponseEntity<String> sendPasswordResetEmail(@RequestBody PasswordForgotRequest request) {
+        authSignUsecase.sendPasswordResetEmail(request.email());
```
`NO_CONTENT` 응답이므로 `void` 또는 `ResponseEntity<Void>`를 사용할 수 있어 보입니다! 👍 

누락된 작업 같아서 제안드립니다.

```suggestion
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public ResponseEntity<Void> sendPasswordResetEmail(@RequestBody PasswordForgotRequest request) {
        authSignUsecase.sendPasswordResetEmail(request.email());
        return ResponseEntity.noContent().build();
```
